src/pytest_bdd/gherkin_parser.py

```diff
@@ -276,11 +316,30 @@ def get_gherkin_document(abs_filename: str = None, encoding: str = "utf-8") -> G
     try:
         gherkin_data = Parser().parse(TokenScanner(feature_file_text))
     except CompositeParserException as e:
-        raise exceptions.FeatureError(
-            e.args[0],
-            e.errors[0].location["line"],
-            linecache.getline(abs_filename, e.errors[0].location["line"]).rstrip("\n"),
-            abs_filename,
-        ) from e
-
+        message = e.args[0]
+        line = e.errors[0].location["line"]
+        line_content = linecache.getline(abs_filename, e.errors[0].location["line"]).rstrip("\n")
+        filename = abs_filename
+        gherkin_error_handler = GherkinParserErrorHandler()
+        gherkin_error_handler(message, line, line_content, filename)
+        # If no patterns matched, raise a generic GherkinParserError
+        raise exceptions.GherkinParseError(f"Unknown parsing error: {message}", line, line_content, filename)
+
+    # At this point, the `gherkin_data` should be valid if no exception was raised
     return GherkinDocument.from_dict(gherkin_data)
+
+
+class GherkinParserErrorHandler:
+    """Parses raw Gherkin parser errors and converts them to human-readable exceptions."""
+
+    def __call__(self, raw_error: str, line: int, line_content: str, filename: str):
+        """Map the error message to a specific exception type and raise it."""
+        # Split the raw_error into individual lines
+        error_lines = raw_error.splitlines()
+
+        # Check each line against all error patterns
+        for error_line in error_lines:
+            for pattern, exception_class, message in ERROR_PATTERNS:
+                if pattern.search(error_line):
+                    # If a match is found, raise the corresponding exception with the formatted message
+                    raise exception_class(message, line, line_content, filename)
```

Comment:
we should `raise ... from e` (the original exception), so that the stacktrace is represented correctly, while now it is represented as
```
...
E   (5:5): expected: #EOF, #TableRow, #DocStringSeparator, #StepLine, #TagLine, #ScenarioLine, #RuleLine, #Comment, #Empty, got 'Background:'

During handling of the above exception, another exception occurred:
...
E   pytest_bdd.exceptions.BackgroundError: BackgroundError: Multiple 'Background' sections detected. Only one 'Background' is allowed per feature.
E   Line number: 5
E   Line:     Background:
E   File: /private/var/folders/dy/_0849zr93fj1td2_07t4ybpr0000gn/T/pytest-of-youtux/pytest-4/test_background_basic0/background.feature
```

Reply:
Fixed and added test